pkg/router/controller/status.go

```diff
@@ -0,0 +1,238 @@
+var nowFn = unversioned.Now
+
+// findOrCreateIngress loops through the router status ingress array looking for an entry
+// that matches name. If there is no entry in the array, it creates one and appends it
+// to the array. If there are multiple entries with that name, the first one is
+// returned and later ones are removed. Changed is returned as true if any part of the
+// array is altered.
+func findOrCreateIngress(route *routeapi.Route, name string) (_ *routeapi.RouteIngress, changed bool) {
+	position := -1
+	updated := make([]routeapi.RouteIngress, 0, len(route.Status.Ingress))
+	for i := range route.Status.Ingress {
+		existing := &route.Status.Ingress[i]
+		if existing.RouterName != name {
+			updated = append(updated, *existing)
+			continue
+		}
+		if position != -1 {
+			changed = true
+			continue
+		}
+		updated = append(updated, *existing)
+		position = i
+	}
+	switch {
+	case position == -1:
+		position = len(route.Status.Ingress)
+		route.Status.Ingress = append(route.Status.Ingress, routeapi.RouteIngress{
+			RouterName: name,
+			Host:       route.Spec.Host,
+		})
+		changed = true
+	case changed:
+		route.Status.Ingress = updated
+	}
+	ingress := &route.Status.Ingress[position]
+	if ingress.Host != route.Spec.Host {
+		ingress.Host = route.Spec.Host
+		changed = true
+	}
+	return ingress, changed
+}
+
+// setIngressCondition records the condition on the ingress, returning true if the ingress was changed and
+// false if no modification was made.
+func setIngressCondition(ingress *routeapi.RouteIngress, condition routeapi.RouteIngressCondition) bool {
+	for _, existing := range ingress.Conditions {
+		//existing.LastTransitionTime = nil
+		if existing == condition {
+			return false
+		}
+	}
+	now := nowFn()
+	condition.LastTransitionTime = &now
+	ingress.Conditions = []routeapi.RouteIngressCondition{condition}
+	return true
+}
+
+func ingressConditionTouched(ingress *routeapi.RouteIngress) *unversioned.Time {
+	var lastTouch *unversioned.Time
+	for _, condition := range ingress.Conditions {
+		if t := condition.LastTransitionTime; t != nil {
+			switch {
+			case lastTouch == nil, t.After(lastTouch.Time):
+				lastTouch = t
+			}
+		}
+	}
+	return lastTouch
+}
+
+// recordIngressConditionFailure updates the matching ingress on the route (or adds a new one) with the specified
+// condition, returning true if the object was modified.
+func recordIngressConditionFailure(route *routeapi.Route, name string, condition routeapi.RouteIngressCondition) (*routeapi.RouteIngress, bool, *unversioned.Time) {
+	for i := range route.Status.Ingress {
+		existing := &route.Status.Ingress[i]
+		if existing.RouterName != name {
+			continue
+		}
+		lastTouch := ingressConditionTouched(existing)
+		return existing, setIngressCondition(existing, condition), lastTouch
+	}
+	route.Status.Ingress = append(route.Status.Ingress, routeapi.RouteIngress{RouterName: name})
+	ingress := &route.Status.Ingress[len(route.Status.Ingress)-1]
```

Comment:
Is it preferable to create a new object inline of a slice append and then retrieve it vs create a new object and add it to a slice?